[dcp] Add ZStandard transformer

mhorowitz opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #143360
* #145528



cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @LucasLLC @MeetVadakkanchery @pradeepfn @ekr0

mhorowitz committed:
[dcp] Add ZStandard transformer

[ghstack-poisoned]

mhorowitz committed:
Update on "[dcp] Add ZStandard transformer"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o LucasLLC MeetVadakkanchery pradeepfn

[ghstack-poisoned]

mhorowitz committed:
Update on "[dcp] Add ZStandard transformer"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o LucasLLC MeetVadakkanchery pradeepfn

[ghstack-poisoned]

mhorowitz committed:
Update on "[dcp] Add ZStandard transformer"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o LucasLLC MeetVadakkanchery pradeepfn

[ghstack-poisoned]

mhorowitz committed:
Update on "[dcp] Add ZStandard transformer"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o LucasLLC MeetVadakkanchery pradeepfn ekr0

[ghstack-poisoned]

mhorowitz committed:
Update on "[dcp] Add ZStandard transformer"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o LucasLLC MeetVadakkanchery pradeepfn ekr0

[ghstack-poisoned]

saumishr commented on test/distributed/checkpoint/test_file_system_checkpoint_cpu.py:
```diff
@@ -20,7 +20,7 @@
     save_state_dict,
 )
-from torch.distributed.checkpoint._extension import Rot13Example
+from torch.distributed.checkpoint._extension import Rot13Example, ZStandard
```
As discussed in the other PR, lets move Rot13Example as a mock in the tests folder.

saumishr commented on test/distributed/checkpoint/test_file_system_checkpoint_cpu.py:
```diff
@@ -164,6 +164,39 @@ def test_read_write_only_tensor(self, thread_count) -> None:
             assert_state_dict_equal(self, state_dict_to_load_to, state_dict_to_save)
 
 
+class TestDistributedStateDictSaveLoadZStandard(TestCase):
```
Are these tests sufficient? I believe in the conversation, team wanted to cover the following data types:
1. blobs save/load
2. non-tensor data save/load
3. Dtensor save/load
4. Re-sharding DTensor save/load 
etc.

saumishr replied:
Also, Multi rank tests? Are those in some other PR?

mhorowitz replied:
Multi-rank tests are in [test/distributed/checkpoint/test_file_system_checkpoint.py](https://github.com/pytorch/pytorch/pull/143360/files#diff-80706d7c648a1e97a397047b0d2e29fccfb710e58fd7ed51665aea6e6c249928) above.

I can add more data type diversity.

mhorowitz committed:
Update on "[dcp] Add ZStandard transformer"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o LucasLLC MeetVadakkanchery pradeepfn ekr0

[ghstack-poisoned]

mhorowitz commented:
@pytorchbot merge

Skylion007 commented:
Any reason to use zstandard over zstd pypi package?

atalman commented:
@pytorchmergebot revert -c nosignal -m "Broke 3.13t builds please test with ciflow/binaries label attached"

malfet commented:
Why do we need zstandard in requirements.txt? And if this is a distributed specific issue, perhaps we should only require it if PyTorch is build with distributed?

mhorowitz committed:
Update on "[dcp] Add ZStandard transformer"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o LucasLLC MeetVadakkanchery pradeepfn ekr0

[ghstack-poisoned]

mhorowitz commented:
> Any reason to use zstandard over zstd pypi package?

zstd has no streaming capability or support for multiple blocks.  zstd's own docs say python-zstandard is "More full-featured and compatible with Zstandard.

I have modified this diff to use pyzstd, which is yet another module wrapping the zstd native library (I don't know why there are so many).  This one appears to be compatible with 3.13t, and, empirically, the artifacts are compatible with python-zstandard.  I'll do some perf testing before I land.

> Why do we need zstandard in requirements.txt? 

https://fb.workplace.com/groups/1144215345733672/posts/1970231166465415/?comment_id=1970283839793481&reply_comment_id=1971945759627289 (sorry for internal link) says that astunparse is a precedent for optional features like this.  Since astunparse was in requirements.txt, I added zstandard (and replaced it with pyzstd) there, too.  If there's a better way to do this, I'm happy to change it.

> And if this is a distributed specific issue, perhaps we should only require it if PyTorch is build with distributed?

It's used by the distributed checkpointing API to add support for checkpoint compression.  If there's a way to require it only if distributed checkpointing is being built, I'm happy to change that, too, but I would need someone to give me an example or teach me how.

mhorowitz committed:
Update on "[dcp] Add ZStandard transformer"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o LucasLLC MeetVadakkanchery pradeepfn ekr0

[ghstack-poisoned]

mhorowitz committed:
Update on "[dcp] Add ZStandard transformer"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o LucasLLC MeetVadakkanchery pradeepfn ekr0

[ghstack-poisoned]

mhorowitz committed:
Update on "[dcp] Add ZStandard transformer"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o LucasLLC MeetVadakkanchery pradeepfn ekr0

[ghstack-poisoned]

albanD commented on .ci/docker/requirements-ci.txt:
```diff
@@ -362,6 +362,7 @@ pwlf==2.2.1 ; python_version >= "3.8"
 # To build PyTorch itself
 astunparse
 PyYAML
+pyzstd
```
This is NOT a build dependency, only an optional runtime dependency?

mhorowitz replied:
correct.

albanD commented on requirements.txt:
```diff
@@ -14,6 +14,7 @@ optree>=0.13.0
 packaging
 psutil
 pyyaml
+pyzstd
```
Why add it here?

mhorowitz replied:
I've removed it for now.  If this causes issues later, we can revisit.

albanD reviewed (commented):
cc @malfet the handling of optional deps is a bit concerning overall. In particular:
- There is no review / consistent handling for user facing messages
- No testing of with / without state
- No testing they don't slip to become a hard dependency (as we always put it there in CI)

mhorowitz committed:
Update on "[dcp] Add ZStandard transformer"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o LucasLLC MeetVadakkanchery pradeepfn ekr0

[ghstack-poisoned]

mhorowitz commented on torch/distributed/checkpoint/_extension.py:
```diff
@@ -1,15 +1,50 @@
+    sys.modules[pyzstd_module_name] = pyzstd
+    pyzstd_spec.loader.exec_module(pyzstd)  # type: ignore[union-attr]
+else:
+    pyzstd = None
```
As discussed, I introduced a helper, in the previous diff.

albanD commented on torch/distributed/checkpoint/_extension.py:
```diff
@@ -1,15 +1,31 @@
+# introduced as collections.abc.Buffer in Python 3.12
+from typing_extensions import Buffer
+
+from torch._utils import try_import
```
You forgot to push that change?

albanD reviewed (approved):
Thanks for the cleanup!

mhorowitz committed:
Update on "[dcp] Add ZStandard transformer"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o LucasLLC MeetVadakkanchery pradeepfn ekr0

[ghstack-poisoned]

mhorowitz committed:
Update on "[dcp] Add ZStandard transformer"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o LucasLLC MeetVadakkanchery pradeepfn ekr0

[ghstack-poisoned]

mhorowitz commented:
@pytorchbot merge

mhorowitz commented:
@pytorchbot merge --force 'PR was landed and reverted.  forcing to land again."

mhorowitz commented:
@pytorchbot merge --force 'PR was landed and reverted. forcing to land again.'

mhorowitz commented:
I had an error (no longer visible) which made it sounds like this wasn't actually merged, which is why I tried again.  Looking at the actual main branch, it was merged, so closing this out.
